src/bijectors/named_bijector.jl

```diff
@@ -0,0 +1,230 @@
+    @assert length(T.parameters) > 0
+    expr = :(x)
+    for i in 1:length(T.parameters)
+        expr = :(cb.bs[$i]($expr))
+    end
+    return expr
+end
+
+function logabsdetjac(cb::NamedComposition, x)
+    y, logjac = forward(cb.bs[1], x)
+    for i = 2:length(cb.bs)
+        res = forward(cb.bs[i], y)
+        y = res.rv
+        logjac += res.logabsdetjac
+    end
+
+    return logjac
+end
+
+@generated function logabsdetjac(cb::NamedComposition{T}, x) where {T<:Tuple}
+    N = length(T.parameters)
+
+    expr = Expr(:block)
+    push!(expr.args, :((y, logjac) = forward(cb.bs[1], x)))
+
+    for i = 2:N - 1
+        temp = gensym(:res)
+        push!(expr.args, :($temp = forward(cb.bs[$i], y)))
+        push!(expr.args, :(y = $temp.rv))
+        push!(expr.args, :(logjac += $temp.logabsdetjac))
+    end
+    # don't need to evaluate the last bijector, only it's `logabsdetjac`
+    push!(expr.args, :(logjac += logabsdetjac(cb.bs[$N], y)))
+
+    push!(expr.args, :(return logjac))
+
+    return expr
+end
+
+
+function forward(cb::NamedComposition, x)
+    rv, logjac = forward(cb.bs[1], x)
+    
+    for t in cb.bs[2:end]
+        res = forward(t, rv)
+        rv = res.rv
+        logjac = res.logabsdetjac + logjac
+    end
+    return (rv=rv, logabsdetjac=logjac)
+end
+
+
+@generated function forward(cb::NamedComposition{T}, x) where {T<:Tuple}
+    expr = Expr(:block)
+    push!(expr.args, :((y, logjac) = forward(cb.bs[1], x)))
+    for i = 2:length(T.parameters)
+        temp = gensym(:temp)
+        push!(expr.args, :($temp = forward(cb.bs[$i], y)))
+        push!(expr.args, :(y = $temp.rv))
+        push!(expr.args, :(logjac += $temp.logabsdetjac))
+    end
+    push!(expr.args, :(return (rv = y, logabsdetjac = logjac)))
+
+    return expr
+end
+
+
+############################
+### `NamedCouplingLayer` ###
+############################
+# TODO: Add ref to `Coupling` or `CouplingLayer` once that's merged.
+"""
+    NamedCoupling{target, deps, F} <: AbstractNamedBijector
+
+Implements a coupling layer for named bijectors.
+
+# Examples
+```julia-repl
+julia> using Bijectors: NamedCoupling, Scale
+
+julia> b = NamedCoupling(:b, (:a, :c), (a, c) -> Scale(a + c))
+NamedCoupling{:b,(:a, :c),var"#3#4"}(var"#3#4"())
+
+julia> x = (a = 1., b = 2., c = 3.);
+
+julia> b(x)
+(a = 1.0, b = 8.0, c = 3.0)
+
+julia> (a = x.a, b = (x.a + x.c) * x.b, c = x.c)
+(a = 1.0, b = 8.0, c = 3.0)
+```
+"""
+struct NamedCoupling{target, deps, F} <: AbstractNamedBijector where {F, target}
+    f::F
+end
+
+NamedCoupling(target, deps, f::F) where {F} = NamedCoupling{target, deps, F}(f)
+function NamedCoupling(::Val{target}, ::Val{deps}, f::F) where {target, deps, F}
+    return NamedCoupling{target, deps, F}(f)
+end
+
+coupling(b::NamedCoupling) = b.f
+# For some reason trying to use the parameteric types doesn't always work
+# so we have to do this weird approach of extracting type and then index `parameters`.
+target(b::NamedCoupling) = typeof(b).parameters[1]
```

Comment:
This is not good. Please use type parameters.

Reply:
Done :+1: